File: tools/patch_semantic.py
import re

with open("src/Semantic.cpp", "r") as f:
    content = f.read()

# Add NamespaceDecl to Semantic.cpp
namespace_code = """        case Parser::NodeType::NamespaceDecl: {
            std::string old_prefix = namespace_prefix;
            namespace_prefix += std::string(pool.get(node.token.data)) + "::";
            for (uint32_t i = 0; i < node.children_count; ++i) {
                check(child_indices[node.children_offset + i]);
            }
            namespace_prefix = old_prefix;
            result = type_table.get_builtin(TypeKind::Void);
            break;
        }"""
content = content.replace("        case Parser::NodeType::FuncDecl: {", namespace_code + "\n        case Parser::NodeType::FuncDecl: {")

# In FuncDecl, prefix the function name with namespace_prefix
content = content.replace("functions[node.token.data] = sig;", "functions[pool.intern(namespace_prefix + std::string(pool.get(node.token.data)))] = sig;")

with open("src/Semantic.cpp", "w") as f:
    f.write(content)

File: tools/patch_semantic.cpp
--- src/Semantic.cpp
+++ src/Semantic.cpp
@@ -326,14 +326,30 @@
             result = type_table.get_builtin(TypeKind::Void);
             break;
         }
-                }
-                return id;
-            };
-
-            if (resolve_alias(ret_type) != resolve_alias(current_func_return_type)) {
-                error(node.token, "Тип возвращаемого значения не совпадает с сигнатурой функции");
-            }
-            result = ret_type;
-            break;
+        case Parser::NodeType::Call: {
+            Parser::NodeId callee_id = child_indices[node.children_offset];
+            Lexer::IdentId func_name = nodes[callee_id].token.data;
+            std::string fname_str = std::string(pool.get(func_name));
+
+            if (fname_str == "печать" || fname_str == "ввод" || fname_str == "выход" || fname_str == "паника" || fname_str == "утверждение") {
+                for (uint32_t i = 1; i < node.children_count; ++i) {
+                    check(child_indices[node.children_offset + i]);
+                }
+                if (fname_str == "ввод") result = type_table.get_builtin(TypeKind::String);
+                else result = type_table.get_builtin(TypeKind::Void);
+                break;
+            }
+
+            if (!functions.count(func_name)) error(node.token, "Вызов неизвестной функции");
+            auto& sig = functions[func_name];
+            if (node.children_count - 1 != sig.param_types.size()) {
+                error(node.token, "Неверное количество аргументов");
+            }
+            for (uint32_t i = 1; i < node.children_count; ++i) {
+                TypeId arg_type = check(child_indices[node.children_offset + i]);
+                if (!types_compatible(arg_type, sig.param_types[i - 1])) error(node.token, "Несоответствие типа аргумента");
+            }
+            result = sig.return_type;
+            break;
+        }
+        case Parser::NodeType::Return: {
+            TypeId ret_type = type_table.get_builtin(TypeKind::Void);
+            if (node.children_count > 0) {
+                ret_type = check(child_indices[node.children_offset]);
+            }
+            if (!types_compatible(ret_type, current_func_return_type)) {
+                error(node.token, "Тип возвращаемого значения не совпадает с сигнатурой функции");
+            }
+            result = ret_type;
+            break;
         }
     default:
